File: VB/Q515371/Models/TreeListVirtualModeHelper.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Web.ASPxTreeList
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Web

Namespace Q515371.Models
	Public Class TreeListVirtualModeHelper
		' Fields...


	   Private Shared nwd As New NorthwindDataContext()
		Public Shared Sub VirtualModeCreateChildren(ByVal e As TreeListVirtualModeCreateChildrenEventArgs)
			Dim parentEmployee As Employee = TryCast(e.NodeObject, Employee)
			If parentEmployee Is Nothing Then
				e.Children = nwd.Employees.Where(Function(empl) empl.ReportsTo Is Nothing).ToList()
			Else
				e.Children = ( _
						From empl In nwd.Employees _
						Where empl.ReportsTo.Equals(parentEmployee.EmployeeID) _
						Select empl).ToList()
			End If
		End Sub
		Public Shared Sub VirtualModeNodeCreating(ByVal e As TreeListVirtualModeNodeCreatingEventArgs)
			Dim empl As Employee = TryCast(e.NodeObject, Employee)

			If empl Is Nothing Then
				Return
			End If
			e.NodeKeyValue = empl.EmployeeID
			Dim childEmployeeCount As Integer = nwd.Employees.Where(Function(x) x.ReportsTo.Equals(empl.EmployeeID)).ToList().Count
			e.IsLeaf = Not(childEmployeeCount > 0)
			e.SetNodeValue("FirstName", empl.FirstName)
			e.SetNodeValue("LastName", empl.LastName)
			e.SetNodeValue("Address", empl.Address)
			e.SetNodeValue("City", empl.City)
			e.SetNodeValue("HireDate", empl.HireDate)
			e.SetNodeValue("BirthDate", empl.BirthDate)
			e.SetNodeValue("ReportsTo", empl.ReportsTo)
			e.SetNodeValue("Title", empl.Title)
		End Sub
		Public Shared Sub MoveNode(ByVal EmployeeID As Integer, ByVal ReportsTo? As Integer)
			Dim newParentID As Integer = Convert.ToInt32(ReportsTo)
			Dim empl As Employee = GetEmployee(EmployeeID)
			If empl.EmployeeID = newParentID Then
				Return
			End If
			If newParentID = 0 Then
				empl.ReportsTo = Nothing
			Else
				empl.ReportsTo = newParentID
			End If
			nwd.SubmitChanges()

		End Sub
		Private Shared Function GetEmployee(ByVal EmployeeID As Integer) As Employee
			Return nwd.Employees.Where(Function(x) x.EmployeeID = EmployeeID).FirstOrDefault()
		End Function
	End Class
End Namespace

File: CS/Q515371/Models/TreeListVirtualModeHelper.cs
using DevExpress.Web.ASPxTreeList;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Web;

namespace Q515371.Models
{
    public class TreeListVirtualModeHelper
    {
        // Fields...
        
            
       static NorthwindDataContext nwd = new NorthwindDataContext();
        public static void VirtualModeCreateChildren(TreeListVirtualModeCreateChildrenEventArgs e) {           
            Employee parentEmployee = e.NodeObject as Employee;
            if (parentEmployee == null)
            {
                e.Children = nwd.Employees.Where(empl=>empl.ReportsTo==null).ToList();
            }
            else {
                e.Children = (from empl in nwd.Employees where empl.ReportsTo == parentEmployee.EmployeeID select empl).ToList();
            }          
        }
        public static void VirtualModeNodeCreating(TreeListVirtualModeNodeCreatingEventArgs e) {
            Employee empl = e.NodeObject as Employee;
            
            if (empl == null)
                return;
            e.NodeKeyValue = empl.EmployeeID;
            int childEmployeeCount = nwd.Employees.Where(x => x.ReportsTo == empl.EmployeeID).ToList().Count;
            e.IsLeaf = !(childEmployeeCount > 0);           
            e.SetNodeValue("FirstName", empl.FirstName);
            e.SetNodeValue("LastName", empl.LastName);
            e.SetNodeValue("Address", empl.Address);
            e.SetNodeValue("City", empl.City);
            e.SetNodeValue("HireDate", empl.HireDate);
            e.SetNodeValue("BirthDate", empl.BirthDate);
            e.SetNodeValue("ReportsTo", empl.ReportsTo);
            e.SetNodeValue("Title", empl.Title);
        }
        public static void MoveNode(int EmployeeID, int? ReportsTo)
        {
            int newParentID = Convert.ToInt32(ReportsTo);
            Employee empl = GetEmployee(EmployeeID);
            if (empl.EmployeeID == newParentID)
                return;
            if (newParentID == 0)
                empl.ReportsTo = null;
            else
                empl.ReportsTo = newParentID;
            nwd.SubmitChanges();
           
        }
        private static Employee GetEmployee(int EmployeeID)
        {
            return nwd.Employees.Where(x => x.EmployeeID == EmployeeID).FirstOrDefault();
        }     
    }
}